Platform detect fails on MacOS (when app is located in temp folders)

## Expected Behavior
Eto-Apps work on MacOS when started from temporary folders by commandline or by code.

## Actual Behavior
Eto-Apps on MacOS believe they should run on Eto.GtkSharp.Platform (on 2.5.0-ci-10148 - should be the latest development version), respectively the message "Could not detect platform" appears (on 2.4.1), when started in temporary folders by commandline or by code. The app will not start.

`Eto.Platform.Detect` returns the wrong (in the current dev branch) platform (and crashes on 2.4.1). But oddly, `EtoEnvironment.Platform.IsMac` returns true and `Type.GetType("Eto.PlatformDetect, Eto.Mac64", false)` (I found this as part of Eto's source code to detect platforms) does not return null, it's AssemblyQualifiedName actually starts with the string `Eto.PlatformDetect, Eto.Mac64`.

## Steps to Reproduce the Problem

  1. Create an empty Eto app named `testapp` in Visual Studio for Mac
  2. Compile it, copy it to `/tmp` or the MacOS temp folder (the long path `Path.GetTempPath()` leads to, see https://apple.stackexchange.com/a/131080 - this will look like `/var/folders/kf/wkq58hz90c98zc7v18bp6nk80000gn/T`)
  3. Execution won't work when using the terminal command `/tmp/testapp.Desktop.app/Contents/MacOS/testapp.Desktop`(yes, the x-bit is set). Execution works when double clicking on the app.
  4. Copy the same testapp.Desktop.app to my user desktop folder by `cp -R /tmp/testapp.Desktop.app /User/myuser/testapp.Desktop.app`.
  5. Execution works now, when using the terminal command `/User/myuser/testapp.Desktop.app/Contents/MacOS/testapp.Desktop` (and also works by double clicking the icon).

<!-- Supplying code that demonstrates the problem expedites a fix (and a response)! -->
## Code that Demonstrates the Problem

```
using System;
using Eto.Forms;
using Eto;

namespace testapp.Desktop
{
    class Program
    {
        [STAThread]
        static void Main(string[] args)
        {
            bool isMac = EtoEnvironment.Platform.IsMac;
            bool is64Bit = EtoEnvironment.Is64BitProcess;
            Type detectType = Type.GetType("Eto.PlatformDetect, Eto.XamMac2", false)
                             ?? Type.GetType("Eto.PlatformDetect, Eto.Mac64", false)
                             ?? Type.GetType("Eto.PlatformDetect, Eto.XamMac", false)
                             ?? Type.GetType("Eto.PlatformDetect, Eto.Mac", false);

            Console.WriteLine($"IsMac: {isMac}, Is64BitProcess: {is64Bit}, detectType.AssemblyQualifiedName {detectType?.AssemblyQualifiedName}");
            new Application(Eto.Platform.Detect).Run(new MainForm());
        }
    }
}
```

### Code output on Eto 2.5.0-ci-10148

When starting from the finder or in non-temp folders: works as expected

When starting from terminal and a temp folder:
```
$ /tmp/testapp.Desktop.app/Contents/MacOS/testapp.Desktop 
IsMac: True, Is64BitProcess: True, detectType.AssemblyQualifiedName Eto.PlatformDetect, Eto.Mac64, Version=2.5.0.0, Culture=neutral, PublicKeyToken=null

Unhandled Exception:
System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
  at GLibrary.Load (Library library) [0x00099] in <f775ff7b2e4a42d992aca2c382879b7c>:0 
  at Gtk.Application..cctor () [0x00000] in <f775ff7b2e4a42d992aca2c382879b7c>:0 
   --- End of inner exception stack trace ---
  at Eto.GtkSharp.Helper.Init () [0x00007] in <e99526d758494eccbbf053da2471a141>:0 
  at Eto.GtkSharp.Forms.ApplicationHandler.Initialize () [0x00006] in <e99526d758494eccbbf053da2471a141>:0 
  at Eto.WidgetHandler`1[TWidget].Eto.Widget.IHandler.Initialize () [0x00000] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Widget.Initialize () [0x0000a] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Widget..ctor () [0x00099] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Forms.Application..ctor (Eto.Forms.Application+InitHelper init) [0x00016] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Forms.Application..ctor (Eto.Platform platform) [0x00007] in <963818933f68425995bee6ad943347da>:0 
  at testapp.Desktop.Program.Main (System.String[] args) [0x00084] in /Users/myuser/Source/testapp/testapp/testapp.Desktop/Program.cs:25 
[ERROR] FATAL UNHANDLED EXCEPTION: System.TypeInitializationException: The type initializer for 'Gtk.Application' threw an exception. ---> System.DllNotFoundException: Gtk
  at GLibrary.Load (Library library) [0x00099] in <f775ff7b2e4a42d992aca2c382879b7c>:0 
  at Gtk.Application..cctor () [0x00000] in <f775ff7b2e4a42d992aca2c382879b7c>:0 
   --- End of inner exception stack trace ---
  at Eto.GtkSharp.Helper.Init () [0x00007] in <e99526d758494eccbbf053da2471a141>:0 
  at Eto.GtkSharp.Forms.ApplicationHandler.Initialize () [0x00006] in <e99526d758494eccbbf053da2471a141>:0 
  at Eto.WidgetHandler`1[TWidget].Eto.Widget.IHandler.Initialize () [0x00000] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Widget.Initialize () [0x0000a] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Widget..ctor () [0x00099] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Forms.Application..ctor (Eto.Forms.Application+InitHelper init) [0x00016] in <963818933f68425995bee6ad943347da>:0 
  at Eto.Forms.Application..ctor (Eto.Platform platform) [0x00007] in <963818933f68425995bee6ad943347da>:0 
  at testapp.Desktop.Program.Main (System.String[] args) [0x00084] in /Users/myuser/Source/testapp/testapp/testapp.Desktop/Program.cs:25 
```

### Code output on Eto 2.4.1

When starting from the finder or in non-temp folders: works as expected

When starting from terminal and a temp folder:
```
$ /tmp/testapp.Desktop.app/Contents/MacOS/testapp.Desktop 
IsMac: True, Is64BitProcess: True, detectType.AssemblyQualifiedName Eto.PlatformDetect, Eto.Mac64, Version=2.4.0.0, Culture=neutral, PublicKeyToken=null

Unhandled Exception:
System.InvalidOperationException: Could not detect platform. Are you missing a platform assembly?
  at Eto.Platform.get_Detect () [0x000d5] in <6db481985cca4510b36c8dbd12852b82>:0 
  at testapp.Desktop.Program.Main (System.String[] args) [0x00073] in /Users/myuser/Source/testapp/testapp/testapp.Desktop/Program.cs:20 
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Could not detect platform. Are you missing a platform assembly?
  at Eto.Platform.get_Detect () [0x000d5] in <6db481985cca4510b36c8dbd12852b82>:0 
  at testapp.Desktop.Program.Main (System.String[] args) [0x00073] in /Users/myuser/Source/testapp/testapp/testapp.Desktop/Program.cs:20 
```

### Code I use to start an app from code

```
using System.Diagnostics;

ProcessStartInfo procInfo = new ProcessStartInfo("/tmp/testapp.Desktop.app/Contents/MacOS/testapp.Desktop")
{
    UseShellExecute = false // this value does not matter
};
Process process = Process.Start(procInfo);
```

This does not work as well, same behaviour as when starting from terminal.

<!-- Please specify the version of Eto and all platforms this can be reproduced in -->
## Specifications

  - Version: 2.5.0-ci-10148, 2.4.1
  - Platform(s): Mac
  - Operating System(s): macOS 10.13.6

--- a/src/Eto.Mac/Platform.cs
+++ b/src/Eto.Mac/Platform.cs
@@ -231,14 +231,14 @@ namespace Eto.Mac
 		{
 			get
 			{
-				var assembly = Assembly.GetEntryAssembly() ?? Assembly.GetExecutingAssembly();
-
-				var assemblyDir = Path.GetDirectoryName(assembly.Location);
-				// location will be empty when embedded via mkbundle, ensure bundlepath is an .app bundle
-				if (string.IsNullOrEmpty(assemblyDir))
-					return NSBundle.MainBundle?.BundlePath.EndsWith(".app", StringComparison.Ordinal) == true;
-
-				return NSBundle.MainBundle != null && assembly.Location.StartsWith(NSBundle.MainBundle.BundlePath, StringComparison.Ordinal);
+				var bundle = NSBundle.MainBundle;
+				if (bundle == null)
+					return false;
+				if (!bundle.BundlePath.EndsWith(".app", StringComparison.Ordinal))
+					return false;
+				if (!bundle.IsLoaded)
+					return false;
+				return true;
 			}
 		}
 

--- a/src/Eto/Forms/Application.cs
+++ b/src/Eto/Forms/Application.cs
@@ -229,7 +229,7 @@ namespace Eto.Forms
 		/// <seealso cref="Platforms"/>
 		/// <param name="platformType">Platform type to initialize this application with</param>
 		public Application(string platformType)
-			: this(Platform.Get(platformType, false))
+			: this(Platform.Get(platformType, false, null))
 		{
 		}
 

--- a/src/Eto/Platform.cs
+++ b/src/Eto/Platform.cs
@@ -387,39 +387,47 @@ namespace Eto
 				if (globalInstance != null)
 					return globalInstance;
 
+				var errors = new List<Exception>();
 				Platform detected = null;
 			
 				if (EtoEnvironment.Platform.IsMac)
 				{
 					if (EtoEnvironment.Is64BitProcess)
-						detected = Platform.Get(Platforms.Mac64, true);
+						detected = Get(Platforms.Mac64, true, errors);
 					if (detected == null)
-						detected = Platform.Get(Platforms.XamMac2, true);
+						detected = Get(Platforms.XamMac2, true, errors);
 					if (detected == null)
-						detected = Platform.Get(Platforms.XamMac, true);
+						detected = Get(Platforms.XamMac, true, errors);
 					if (detected == null)
-						detected = Platform.Get(Platforms.Mac, true);
+						detected = Get(Platforms.Mac, true, errors);
 				}
 				else if (EtoEnvironment.Platform.IsWindows)
 				{
-					detected = Platform.Get(Platforms.Wpf, true);
+					detected = Get(Platforms.Wpf, true, errors);
 					if (detected == null)
-						detected = Platform.Get(Platforms.WinForms, true);
+						detected = Get(Platforms.WinForms, true, errors);
 				}
 
 				if (detected == null && EtoEnvironment.Platform.IsUnix)
 				{
-					detected = Platform.Get(Platforms.Gtk, true);
+					detected = Get(Platforms.Gtk, true, errors);
 #pragma warning disable CS0618
 					if (detected == null)
-						detected = Platform.Get(Platforms.Gtk3, true);
+						detected = Get(Platforms.Gtk3, true, errors);
 					if (detected == null)
-						detected = Platform.Get(Platforms.Gtk2, true);
+						detected = Get(Platforms.Gtk2, true, errors);
 #pragma warning restore CS0618
 				}
-				
+
 				if (detected == null)
-					throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, "Could not detect platform. Are you missing a platform assembly?"));
+				{
+					var message = "Could not detect platform. Are you missing a platform assembly?";
+					if (errors.Count > 1)
+						throw new AggregateException(message, errors);
+					if (errors.Count == 1)
+						throw new InvalidOperationException(message, errors[0]);
+					throw new InvalidOperationException(message);
+				}
 					
 				Initialize(detected);
 				return globalInstance;
@@ -456,38 +464,44 @@ namespace Eto
 		/// <returns>An instance of a Generator of the specified type, or null if it cannot be loaded</returns>
 		public static Platform Get(string generatorType)
 		{
-			return Get(generatorType, true);
+			return Get(generatorType, true, null);
 		}
 
-		internal static Platform Get(string platformType, bool allowNull)
+		internal static Platform Get(string platformType, bool allowNull, List<Exception> errors)
 		{
 			Type type = Type.GetType(platformType);
 			if (type == null)
 			{
 				if (allowNull)
 					return null;
-				throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, "Platform not found. Are you missing the platform assembly?"));
+				throw new InvalidOperationException($"Platform type '{platformType}' was not found. Are you missing the platform assembly?");
 			}
 			try
 			{
 				var platform = (Platform)Activator.CreateInstance(type);
-				if (!platform.IsValid)
-				{
-					var message = string.Format("Platform type {0} was loaded but is not valid in the current context.  E.g. Mac platforms require to be in an .app bundle to run", platformType);
-					if (allowNull)
-						Debug.WriteLine(message);
-					else
-						throw new InvalidOperationException(string.Format(CultureInfo.CurrentCulture, message));
-					return null;
-				}
-				return platform;
+				if (platform.IsValid)
+					return platform;
+
+				var message = $"Platform type '{platformType}' was loaded but is not valid in the current context.  E.g. Mac platforms require to be in an .app bundle to run";
+				var error = new InvalidOperationException(message);
+				errors?.Add(error);
+				if (!allowNull)
+					throw error;
+
+				Debug.WriteLine(message);
+				return null;
 			}
 			catch (Exception ex)
 			{
-				Debug.WriteLine(string.Format("Error creating instance of platform type '{0}'\n{1}", platformType, ex));
-				if (allowNull)
-					return null;
-				throw;
+				var message = $"Error creating instance of platform type '{platformType}'";
+				var error = new InvalidOperationException(message, ex);
+				errors?.Add(error);
+				if (!allowNull)
+					throw error;
+
+				Debug.WriteLine(message);
+				Debug.WriteLine(ex);
+				return null;
 			}
 		}
 

